store/sqlstore/post_store.go

```diff
@@ -616,10 +659,105 @@ func (s *SqlPostStore) PermanentDeleteByChannel(channelId string) error {
 	return nil
 }
 
+func (s *SqlPostStore) prepareThreadedResponse(posts []*postWithExtra, extended, reversed bool) (*model.PostList, error) {
+	list := model.NewPostList()
+	var userIds []string
+	userIdMap := map[string]bool{}
+	for _, thread := range posts {
+		for _, participantId := range thread.ThreadParticipants {
+			if _, ok := userIdMap[participantId]; !ok {
+				userIdMap[participantId] = true
+				userIds = append(userIds, participantId)
+			}
+		}
+	}
+	var users []*model.User
+	if extended {
+		var err error
+		users, err = s.User().GetProfileByIds(userIds, &store.UserGetByIdsOpts{}, true)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		for _, userId := range userIds {
+			users = append(users, &model.User{Id: userId})
+		}
+	}
+	processPost := func(p *postWithExtra) error {
+		p.Post.ReplyCount = p.ThreadReplyCount
+		for _, th := range p.ThreadParticipants {
+			var participant *model.User
+			for _, u := range users {
+				if u.Id == th {
+					participant = u
+					break
+				}
+			}
+			if participant == nil {
+				return errors.New("cannot find thread participant with id=" + th)
+			}
+			p.Post.Participants = append(p.Post.Participants, participant)
+		}
+		return nil
+	}
+	// We need to flip the order if we selected backwards
+	if !reversed {
+		for _, p := range posts {
+			if err := processPost(p); err != nil {
+				return nil, err
+			}
+			list.AddPost(&p.Post)
+			list.AddOrder(p.Id)
+		}
+	} else {
+		l := len(posts)
+		for i := range posts {
+			if err := processPost(posts[l-i-1]); err != nil {
+				return nil, err
+			}
+			list.AddPost(&posts[l-i-1].Post)
+			list.AddOrder(posts[l-i-1].Id)
+		}
+	}
```

Comment:
Maybe we could simplify a bit here by having one loop only and setting the index accordingly.